Mathlib/Data/Finsupp/Defs.lean

```diff
@@ -1330,4 +1330,10 @@ theorem update_eq_sub_add_single [AddGroup G] (f : α →₀ G) (a : α) (b : G)
   rw [update_eq_erase_add_single, erase_eq_sub_single]
 #align finsupp.update_eq_sub_add_single Finsupp.update_eq_sub_add_single
 
+variable (α) in
+theorem finite_setOf_bounded [Finite α] (N : ℕ) : Finite {n : α →₀ ℕ | ∀ (i : α), n i ≤ N} := by
```

Comment:
This is true for any upper bound `g : α → ℕ`, not necessary a constant. BTW, do we have a similar lemma about functions (or a generic `Set.pi univ` combinator)? If yes, then we can use `Set.Finite.preimage`.

Reply:
This is basically docs#Set.Finite.pi combined with finiteness of intervals. I don't think we should even have it as a lemma. It can be inlined as `(Set.Finite.pi fun _ ↦ .finite_le_nat).preimage $ Funlike.coe_injective.injOn _` (untested)

Reply:
[Set.Finite.pi](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Finite.html#Set.Finite.pi) would give you an extra `univ` as in `∀ i ∈ univ, n i ≤ N`, which can't be easily inlined as far as I know. Moreover this lemma is used to prove that `{s : σ →₀ ℕ | s.sum (fun _ e ↦ e) ≤ N}` is also finite, by virtue of being a subset.

Reply:
This is all handled by the `preimage` step. This is where you get flexibility.

Reply:
> This is basically `Set.Finite.pi` combined with finiteness of intervals. I don't think we should even have it as a lemma. It can be inlined as `(Set.Finite.pi fun _ ↦ .finite_le_nat).preimage $ Funlike.coe_injective.injOn _` (untested)

I've tried to make your suggestion work, but by the time it compiled, it ended up worse than what I had in the first place. Feel free to push a better proof if you have one.

Reply:
Done